dissect/volume/md/stream.py

```diff
@@ -0,0 +1,429 @@
+from __future__ import annotations
+
+import operator
+from typing import TYPE_CHECKING, NamedTuple, Optional
+
+from dissect.util.stream import AlignedStream, MappingStream
+
+from dissect.volume.exceptions import MDError
+from dissect.volume.md.c_md import SECTOR_SIZE, c_md
+
+if TYPE_CHECKING:
+    from dissect.volume.md.md import MD, Device
+
+
+class LinearStream(MappingStream):
+    """Implements a stream on a linear RAID set."""
+
+    def __init__(self, md: MD):
+        super().__init__()
+        self.md = md
+
+        offset = 0
+        for device in md.devices:
+            # Ignore special device roles for now
+            if device.raid_disk is None:
+                continue
+
+            device_size = device.data_size * SECTOR_SIZE
+            self.add(offset, device_size, device.open(), 0)
+            offset += device_size
+
+
+class Zone(NamedTuple):
+    zone_end: int
+    dev_start: int
+    devices: list[Device]
+
+
+class RAID0Stream(AlignedStream):
+    """Implements a stream on a RAID0 set."""
+
+    def __init__(self, md: MD):
+        self.md = md
+        self.devices = {dev.raid_disk: dev for dev in self.md.devices if dev.raid_disk is not None}
+        if len(self.devices) != md.raid_disks:
+            raise MDError(f"Missing disks in RAID0 set {self.md.uuid} ({self.md.name})")
+
+        # Determine how many strip zones we need to construct
+        # If a RAID0 set consists of devices with different sizes, additional strip zones
+        # may exist on the larger devices but not on the smaller ones
+        # Reference: create_strip_zones
+        devices = sorted(self.devices.values(), key=operator.attrgetter("raid_disk"))
+        rounded_sectors = {}
+
+        num_strip_zones = 0
+        for dev1 in devices:
+            rounded_sectors[dev1] = (dev1.sectors // dev1.chunk_sectors) * dev1.chunk_sectors
+
+            has_same_sectors = False
+            # Check if dev1 is unequal in size to the sizes of any of the previous devices
+            # If so, this means an extra strip zone is present
+            for dev2 in devices:
+                if dev1.dev_number == dev2.dev_number:
+                    break
+
+                if rounded_sectors[dev1] == dev2.sectors:
+                    has_same_sectors = True
+                    break
+
+            if not has_same_sectors:
+                num_strip_zones += 1
+
+        # Determine the smallest device and calculate the overal volume size at the same time
+        size = 0
+        smallest = None
+        for dev in devices:
+            if not smallest or rounded_sectors[dev] < rounded_sectors[smallest]:
+                smallest = dev
+
+            # Calculate a rounded up size of the RAID0 set in sectors
+            size += rounded_sectors[dev] & ~(self.md.chunk_sectors - 1)
```

Comment:
If I calculate correct, this is always rounded _down_ to the nearest `md.chunk_sectors`.

`rounded_sectors` is already rounded down to the nearest `device.chunk_sectors`. Say `device.chunk_sectors = 8`, for a number of different  `rounded_sectors` with an `md.chunck_sectors = 16`, this gives:

```python
8 & ~(16 - 1) == 0
16 & ~(16 - 1) == 16
24 & ~(16 - 1) == 16
32 & ~(16 - 1) == 32
```

Reply:
I don't think this is relevant anymore as this file has been superseded by `raid/stream.py` where this was removed.